modules/core/src/smithy4s/Blob.scala

```diff
@@ -18,17 +18,93 @@ package smithy4s
 
 import java.nio.ByteBuffer
 import java.util.Base64
+import scala.collection.immutable.Queue
 import java.nio.charset.StandardCharsets
+import java.io.OutputStream
 
+// scalafmt: {maxColumn = 120}
 sealed trait Blob {
-  def toArray: Array[Byte]
-  def buffer: ByteBuffer
-  def isEmpty: Boolean
+
+  def apply(i: Int): Byte
   def size: Int
-  def sameBytesAs(other: Blob): Boolean
-  final def toArrayBlob: Blob = Blob(toArray)
+
+  def foreach(f: Byte => Unit) = {
+    var i = 0
+    while (i < size) { f(apply(i)); i += 1 }
+  }
+
+  def foreachWithIndex(f: (Byte, Int) => Unit) = {
+    var i = 0
+    while (i < size) { f(apply(i), i); i += 1 }
+  }
+
+  def toArray: Array[Byte] = {
+    val result = Array.ofDim[Byte](size)
+    foreachWithIndex((b, i) => result(i) = b)
+    result
+  }
+
+  def toArrayUnsafe: Array[Byte] = toArray
```

Comment:
Another useful one to have is `toArraySliceBlob: ArraySliceBlob`. This would make it extremely easy to convert to/from `fs2.Chunk.ArraySlice`. It also provides direct (unsafe) access to the array without copying if its a view and without having to grapple with the confusing `ByteBuffer` API.